torchrl/modules/planners/mpc.py

```diff
@@ -0,0 +1,186 @@
+    At the end, we recover the best action which is the one that maximizes the reward given a planning horizon.
+
+    Args:
+        env (Environment): The environment to perform the planning step on (Can be ModelBasedEnv or EnvBase).
+        planning_horizon (int): The number of steps to perform the planning step for.
+        optim_steps (int): The number of steps to perform the MPC planning step for.
+        num_candidates (int): The number of candidates to sample from the Gaussian distribution.
+        num_top_k_candidates (int): The number of top candidates to use to update the mean and standard deviation of the Gaussian distribution.
+        reward_key (str): The key in the TensorDict to use to retrieve the reward.
+        action_key (str): The key in the TensorDict to use to store the action.
+
+    Returns:
+        TensorDict: The TensorDict with the action added.
+    """
+
+    def __init__(
+        self,
+        env: EnvBase,
+        planning_horizon: int,
+        optim_steps: int,
+        num_candidates: int,
+        num_top_k_candidates: int,
+        reward_key: str = "reward",
+        action_key: str = "action",
+    ):
+        super().__init__(env, action_key)
+        self.planning_horizon = planning_horizon
+        self.optim_steps = optim_steps
+        self.num_candidates = num_candidates
+        self.num_top_k_candidates = num_top_k_candidates
+        self.reward_key = reward_key
+
+    def planning(self, tensordict: TensorDictBase) -> torch.Tensor:
+        batch_size = tensordict.batch_size
+        expanded_original_td = (
+            tensordict.clone().expand(*batch_size, self.num_candidates).flatten()
+        )
+        flatten_batch_size = batch_size.numel()
+        actions_means = torch.zeros(
+            flatten_batch_size,
+            1,
+            self.planning_horizon,
+            *self.action_spec.shape,
+            device=tensordict.device,
+        )
+        actions_stds = torch.ones(
+            flatten_batch_size,
+            1,
+            self.planning_horizon,
+            *self.action_spec.shape,
+            device=tensordict.device,
+        )
+        for _ in range(self.optim_steps):
+            actions = actions_means + actions_stds * torch.randn(
+                flatten_batch_size,
+                self.num_candidates,
+                self.planning_horizon,
+                *self.action_spec.shape,
+                device=tensordict.device,
+            )
+            actions = actions.view(
+                flatten_batch_size * self.num_candidates,
+                self.planning_horizon,
+                *self.action_spec.shape,
+            )
+            optim_td = expanded_original_td.clone()
+            policy = PrecomputedActionsSequentialSetter(actions)
+            optim_td = self.env.rollout(
+                max_steps=self.planning_horizon,
+                policy=policy,
+                auto_reset=False,
```

Comment:
don't we want to reset the env?
Otherwise we'll keep on interating like nothing happened and the actions at time t won't match from one iter of the loop to the other

Reply:
We don 't want to reset it since we want the rollout to happen at the given observation state that we are optimizing for. However you are right that after one optim step we need to reset it to the state at the time we optimize for... This is a complex one. It is trivial for MB env since they are stateless but this is a big problem for gym envs that are stateful... Maybe we should just make a copy of the env and reset env to that copy after the rollout?

Reply:
> We don 't want to reset it since we want the rollout to happen at the given observation state that we are optimizing for.

So what I get from your comment is this (correct me if i'm wrong)
you have an env at state _T_, and you want to plan the action from _T_ to _T_ + _t_.
To do that you run this optimization loop several times.
If you reset, you're set back at state _0_ < _T_, which you don't want. Resetting is not an option, but you'd like to reset at time _T_, right?

This can't be done with gym and it's not a use case that is supported by the library. The assumption of gym envs is that you can't reset states where you want them to be (at least not natively). You should not do that, and each env will have its own way of preventing or allowing you to do that, i.e. there is no generic way of setting a state in gym.

What we should do is raise an error if the environment is stateful. This can be checked by assessing if the `input_spec` keys match the `observation_spec` keys (without the `"next_" prefixes). If they don't you can assume that the env is stateful (bc it outputs stuff that was not its input).

Or we could add a flag `is_stateful` to the envs, but if you're doing that I would suggest to write that in a separate PR.

Reply:
> So what I get from your comment is this (correct me if i'm wrong)
you have an env at state T, and you want to plan the action from T to T + t.
To do that you run this optimization loop several times.
If you reset, you're set back at state 0 < T, which you don't want. Resetting is not an option, but you'd like to reset at time T, right?

Its almost that, we only optimize for action T+1 but we look in the future until action T+t

Ok i see so, we should accept that this planning algorithm won't work on gym like envs then. I don't really get the `input_spec` `observation_spec` trick. Like for example in MB_envs the keys won't match (input spec will have actions and observation will not!

Reply:
yeah that's a detail, you can remove action (it'll always be there)
In general that will work. If it doesn't someome will raise an issue and we'll deal with it